### PR TITLE
Add full functionality for handling schedules

### LINKED ISSUE #30: Support schedule configuration like CronJob
## What you want to add
Add schedule node count schedule configuration like this
```
spec:
  scaleTargetRef:
    projectId: your-gcp-project-id
    instanceId: your-spanner-instance-id
  minNodes: 1
  maxNodes: 4
  cron:
    schedule: "*/1 * * * *"
    period: 1H
    minNodes: 4
```
## Why this is needed
We sometimes want to increase node num before starting a batch job.

### PR DESCRIPTION
**Pull request by rustycl0ck**
## What this PR does / Why we need it

Contents of this PR:
- Update the type of `status.CurrentlyActiveSchedules` field of `SpannerAutoscaler` resource for handling schedules
- Add and remove some fields in the status:
  - remove fields which just repeat the processing units compute capacity as nodes
  - add fields for showing current `min` and `max` compute capacity range, based on active schedules
- Fix test cases for handling the above status field changes
- Auto generate CRD schema for the above changes
- Add one more sample `SpannerAutoscaleSchedule` for example
- Implement `internal/scheduler` package. It's main functions are:
  - iterate over `status.CurrentlyActiveSchedules` of `SpannerAutoscaler` resources and remove expired schedules
  - define/implement `Job` interface from `cron` package, for updating `status.CurrentlyActiveSchedules` on each cron schedule trigger
- Fix `SpannerAutoscaleScheduleReconciler` (as well as `syncer` package) to retry status updates whenever k8s returns conflict error: `the object has been modified; please apply your changes to the latest version and try again`
- Update `SpannerAutoscalerReconciler` logic to:
  - create a `scheduler` if one does not exist
  - create a [`cron`](https://pkg.go.dev/github.com/robfig/cron/v3#Cron) for each `SpannerAutoscaler` (only if there is a schedule registered for it)
  - create a cron job for each schedule which is registered with a `SpannerAutoscaler` (this job is responsible for populating the `status.CurrentlyActiveSchedules` of the autoscaler resource on each trigger)
  - remove the `cron` from memory whenever a `SpannerAutoscaler` resource is deleted
  - remove the schedule from `status.CurrentlyActiveSchedules` whenever a schedule is deleted
  - calculate and update compute capacity by including active schedules too

## Which issue(s) this PR fixes

Fixes #30 

**NOTE:** There are some `TODO` comments to handle edge cases and to improve functionality. These will be fixed in the next iteration.

### CODE REVIEW

**Comment by tkuchiki** [comment]
Next, I will test on the `kind` cluster.

File: controllers/spannerautoscaler_controller.go [hunk 1]
```diff
@@ -179,35 +185,46 @@ func NewSpannerAutoscalerReconciler(
...
+				syncer.Stop()
+				r.mu.Lock()
+				delete(r.syncers, nnsa)
+				r.mu.Unlock()
+				log.Info("stopped syncer")
+			}
+
+			if cronExists {
+				cron.Stop()
+				delete(r.crons, nnsa)
```
[the remark this answers is not in the document]
Comment (rustycl0ck) [reply controllers/spannerautoscaler_controller.go:218]: Thanks, applied the suggestion now!

File: internal/scheduler/scheduler.go [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+}
+
+func (j Job) Run() {
+	j.Log.V(1).Info("Cron Job is now run", "now", metav1.Now())
+	var (
+		sa  spannerv1beta1.SpannerAutoscaler
+		sas spannerv1beta1.SpannerAutoscaleSchedule
+	)
+
+	ctx := context.TODO()
```
[the remark this answers is not in the document]
Comment (rustycl0ck) [reply internal/scheduler/scheduler.go]: Thanks, applied the suggestion now!

File: api/v1beta1/spannerautoscaleschedule_types.go [hunk 3]
```diff
@@ -39,6 +39,8 @@ type SpannerAutoscaleScheduleStatus struct{}
 
 //+kubebuilder:object:root=true
 //+kubebuilder:subresource:status
+//+kubebuilder:printcolumn:name="Cron",type="string",JSONPath=".spec.schedule.cron"
+//+kubebuilder:printcolumn:name="Duration",type="string",JSONPath=".spec.schedule.duration"
```
Comment (tkuchiki) [inline api/v1beta1/spannerautoscaleschedule_types.go:44]:
```suggestion
//+kubebuilder:printcolumn:name="Additional PU",type="integer",JSONPath=".spec.additionalProcessingUnits"
```

It seems better to add `Additional Processing Units`.

Comment (tkuchiki) [reply api/v1beta1/spannerautoscaleschedule_types.go:44]:
```console
$ kubectl get spannerautoscaleschedule -n test
NAME                              CRON           DURATION   ADDITIONAL PU
another-sample                    */25 * * * *   15m30s     300
spannerautoscaleschedule-sample   0 */3 * * *    2h10m      3000
```

**Comment by tkuchiki** [comment]
```console
$ kubectl get spannerautoscaleschedule -n test
NAME                              CRON           DURATION   ADDITIONAL PU
another-sample                    */25 * * * *   15m30s     300
spannerautoscaleschedule-sample   0 */3 * * *    2h10m      300
```

```console
$ kubectl get spannerautoscaler -n test
NAME                            PROJECT ID          INSTANCE ID                    CURRENT PUS   DESIRED PUS   DESIRED MIN PUS   DESIRED MAX PUS   TARGET CPU   CURR
ENT CPU   AGE
spannerautoscaler-sample-beta   merpay-alcazar-jp   merpay-alcazar-autoscaler-pu   400           100           100               500               30           2
          88m

$ kubectl get spannerautoscaler -n test
NAME                            PROJECT ID          INSTANCE ID                    CURRENT PUS   DESIRED PUS   DESIRED MIN PUS   DESIRED MAX PUS   TARGET CPU   CURR
ENT CPU   AGE
spannerautoscaler-sample-beta   merpay-alcazar-jp   merpay-alcazar-autoscaler-pu   400           400           400               800               30           3
          99m
```

**Comment by tkuchiki** [comment]
```console
$ kubectl get spannerautoscaleschedule -n test
NAME                              CRON           DURATION   ADDITIONAL PU
another-sample                    */25 * * * *   15m30s     300
spannerautoscaleschedule-sample   0 */3 * * *    2h10m      300
```

```console
$ kubectl get spannerautoscaler -n test
NAME                            PROJECT ID          INSTANCE ID     CURRENT PUS   DESIRED PUS   DESIRED MIN PUS   DESIRED MAX PUS   TARGET CPU   CURR
ENT CPU   AGE
spannerautoscaler-sample-beta   test-project        test-instance   400           100           100               500               30           2
          88m

$ kubectl get spannerautoscaler -n test
NAME                            PROJECT ID          INSTANCE ID     CURRENT PUS   DESIRED PUS   DESIRED MIN PUS   DESIRED MAX PUS   TARGET CPU   CURR
ENT CPU   AGE
spannerautoscaler-sample-beta   test-project        test-instance   400           400           400               800               30           3
          99m
```

It seems good 👍

**Comment by tkuchiki** [comment]
I tried additionalPU setting  to 3000, I faced the following error:

```
failed to update spanner instance","autoscaler":"test/spannerautoscaler-sample-beta","error":"rpc error: code = InvalidArgument desc = Invalid UpdateInstance request.\nerror details: name = BadRequest field = instance.processing_units desc = Processing units must be specified in increments of 100 up to 1000 processing units, or increments of 1000 above 1000 processing units
```

This means that when minPU = 100 + 3,000 PU = desiredMinPU = 3,100 PU, which was a BadRequest because it is not a number divisible by 1000.

So we have two possible approaches:

1. Round up Desired Processing Units
2. Create a Validation for Processing Units

I think `2` is difficult to implement, and if a user has multiple `SpannerAutoscalerSchedule` resources, it seems to be difficult to use because it must be considered so that multiple additionalPUs are added without causing a validation error.

For this reason, I recommend `1`.
BTW, the reason for rounding up instead of rounding down is because I believe it is safer for the load.

Please let me know if you have a better approach!

**Comment by rustycl0ck** [comment]
> I tried additionalPU setting to 3000, I faced the following error:
> 
> ```
> failed to update spanner instance","autoscaler":"test/spannerautoscaler-sample-beta","error":"rpc error: code = InvalidArgument desc = Invalid UpdateInstance request.\nerror details: name = BadRequest field = instance.processing_units desc = Processing units must be specified in increments of 100 up to 1000 processing units, or increments of 1000 above 1000 processing units
> ```
> 
> This means that when maxPU = 500 + 3,000 PU = desiredMaxPU = 3,500 PU, which was a BadRequest because it is not a number divisible by 1000.
> 
> So I think we have two possible approaches:
> 
>     1. Round up Desired Processing Units
> 
>     2. Create a Validation for Processing Units
> 
> 
> I think `2` is difficult to implement, and if a user has multiple `SpannerAutoscalerSchedule` resources, it seems to be difficult to use because it must be considered so that multiple additionalPUs are added without causing a validation error.
> 
> For this reason, I recommend `1`. BTW, the reason for rounding up instead of rounding down is because I believe it is safer for the load.
> 
> Please let me know if you have a better approach!

@tkuchiki san,
Approach number 1, as suggested by you, has been implemented in a later PR [here](https://github.com/mercari/spanner-autoscaler/pull/81/commits/4b7a983a6f0ae3ed94c1eb754d95dbad8b11e775#diff-e627b3d00babe38223bf984ec81c1a175b03c7ac34a19106f98ab76af3265e5dR572-R579). Please confirm if this is acceptable. (I did not fix it here orignally because this PR was already too large, and I didn't want to create more confusion by continuously making changes here.)

**Comment by tkuchiki** [comment]
Thank you for confirming.
I checked the [PR#81](https://github.com/mercari/spanner-autoscaler/pull/81/commits/4b7a983a6f0ae3ed94c1eb754d95dbad8b11e775#diff-e627b3d00babe38223bf984ec81c1a175b03c7ac34a19106f98ab76af3265e5dR572-R579), LGTM!

**Review by tkuchiki — APPROVED** [review]
